src/QsCompiler/Transformations/CallGraphWalker.cs

```diff
@@ -0,0 +1,394 @@
+            //            if (accum.Add((k.Item1, maxDepType)))
+            //            {
+            //                // ToDo: this won't work once Type specialization are implemented
+            //                var noTypeParams = new CallGraphNode { CallableName = k.Item1.CallableName, Kind = k.Item1.Kind, TypeArgs = QsNullable<ImmutableArray<ResolvedType>>.Null };
+            //                WalkDependencyTree(noTypeParams, accum, maxDepType);
+            //            }
+            //        }
+            //    }
+            //
+            //    return accum;
+            //}
+            //
+            //return WalkDependencyTree(callerSpec, new HashSet<(CallGraphNode, DependencyType)>(), DependencyType.NoTypeParameters).ToImmutableArray();
+        }
+
+        /// <summary>
+        /// Returns all specializations that are used directly or indirectly within the given caller,
+        /// whether they are called, partially applied, or assigned. Each key in the returned dictionary
+        /// represents a specialization that is used by the caller. Each value in the dictionary is an
+        /// array of edges representing all the different ways the given caller specialization took a
+        /// dependency on the specialization represented by the associated key.
+        /// </summary>
+        public Dictionary<CallGraphNode, ImmutableArray<CallGraphEdge>> GetAllDependencies(QsSpecialization callerSpec) =>
+            GetAllDependencies(new CallGraphNode { CallableName = callerSpec.Parent, Kind = callerSpec.Kind, TypeArgs = RemovePositionFromTypeArgs(callerSpec.TypeArguments) });
+
+        /// <summary>
+        /// Finds and returns a list of all cycles in the call graph, each one being represented by an array of nodes.
+        /// To get the edges between the nodes of a given cycle, use the GetDirectDependencies method.
+        /// </summary>
+        public List<ImmutableArray<CallGraphNode>> GetCallCycles()
+        {
+            // ToDo: need to implement a robust algorithm for finding cycles
+            throw new NotImplementedException();
+        }
+    }
+
+    /// <summary>
+    /// This transformation walks through the compilation without changing it, building up a call graph as it does.
+    /// This call graph is then returned to the user.
+    /// </summary>
+    public static class BuildCallGraph
+    {
+        public static CallGraph Apply(QsCompilation compilation)
+        {
+            var walker = new BuildGraph();
+            foreach (var ns in compilation.Namespaces)
+            {
+                walker.Namespaces.OnNamespace(ns);
+            }
+            return walker.SharedState.graph;
+        }
+
+        private class BuildGraph : SyntaxTreeTransformation<TransformationState>
+        {
+            public BuildGraph() : base(new TransformationState())
+            {
+                this.Namespaces = new NamespaceTransformation(this);
```

Comment:
```suggestion
                this.Namespaces = new NamespaceTransformation(this, TransformationOptions.NoRebuild);
```

Reply:
Expression Kind and Namespace sub-transformations don't take another parameter. This 'No Rebuild' behavior is hard-coded into them because they are specific to this walker.